File: app/src/main/java/com/example/core/challenge/ServerScoringValidator.kt
package com.example.core.challenge

import com.example.core.scoring.QuestionScoreResult
import com.example.core.scoring.ScoringEngine
import com.example.core.scoring.SpeedFeedbackType
import com.example.data.model.AnswerSubmissionEvent
import com.example.data.model.Question

sealed class AnswerValidationResult {
    data class Valid(val scoreResult: QuestionScoreResult) : AnswerValidationResult()
    data class Rejected(val reason: AntiCheatViolation) : AnswerValidationResult()
}

enum class AntiCheatViolation {
    CHALLENGE_EXPIRED,
    DUPLICATE_QUESTION_SUBMISSION,
    SEQUENCE_MISMATCH,
    INVALID_QUESTION_ID,
    IMPOSSIBLE_RESPONSE_TIME,
    REPLAY_DETECTED
}

/**
 * Server-authoritative Scoring & Anti-Cheat Validator.
 * Prevents client-side manipulation of score, sequence, and timestamps.
 */
class ServerScoringValidator(
    val challengeId: String,
    val startTimeMillis: Long,
    val endTimeMillis: Long,
    private val officialSequence: List<Question>
) {
    private val answeredQuestionIds = mutableSetOf<String>()
    private val processedEventIds = mutableSetOf<String>()
    var verifiedScore: Double = 0.0
        private set
    var currentCombo: Int = 0
        private set
    var bestCombo: Int = 0
        private set
    var totalAnswered: Int = 0
        private set
    var correctCount: Int = 0
        private set
    var fastAnswersCount: Int = 0
        private set
    private val responseTimes = mutableListOf<Double>()

    /**
     * Authoritatively validates and scores an incoming AnswerSubmissionEvent.
     */
    fun validateAndScore(
        event: AnswerSubmissionEvent,
        serverTimestamp: Long = System.currentTimeMillis()
    ): AnswerValidationResult {
        // 1. Check replay attack
        if (processedEventIds.contains(event.eventId)) {
            return AnswerValidationResult.Rejected(AntiCheatViolation.REPLAY_DETECTED)
        }
        processedEventIds.add(event.eventId)

        // 2. Check challenge expiration
        if (serverTimestamp > endTimeMillis) {
            return AnswerValidationResult.Rejected(AntiCheatViolation.CHALLENGE_EXPIRED)
        }

        // 3. Check duplicate question submission
        if (answeredQuestionIds.contains(event.questionId)) {
            return AnswerValidationResult.Rejected(AntiCheatViolation.DUPLICATE_QUESTION_SUBMISSION)
        }

        // 4. Validate sequence position
        if (event.sequencePosition < 0 || event.sequencePosition >= officialSequence.size) {
            return AnswerValidationResult.Rejected(AntiCheatViolation.SEQUENCE_MISMATCH)
        }
        val expectedQuestion = officialSequence[event.sequencePosition]
        if (expectedQuestion.questionId != event.questionId) {
            return AnswerValidationResult.Rejected(AntiCheatViolation.SEQUENCE_MISMATCH)
        }

        // 5. Check impossible response time (< 50ms human reflex floor)
        if (event.responseTimeMs < 50) {
            return AnswerValidationResult.Rejected(AntiCheatViolation.IMPOSSIBLE_RESPONSE_TIME)
        }

        // Mark answered
        answeredQuestionIds.add(event.questionId)
        totalAnswered++

        val responseSeconds = event.responseTimeMs / 1000.0
        responseTimes.add(responseSeconds)

        val isCorrect = event.selectedAnswer == expectedQuestion.correctAnswer

        val scoreResult = ScoringEngine.scoreAnswer(
            isCorrect = isCorrect,
            responseTimeSeconds = responseSeconds,
            previousCombo = currentCombo,
            previousBestCombo = bestCombo
        )

        // Update server authoritative state
        verifiedScore += scoreResult.totalQuestionScore
        currentCombo = scoreResult.currentCombo
        bestCombo = scoreResult.bestCombo
        if (isCorrect) {
            correctCount++
            if (responseSeconds <= ScoringEngine.SPEED_WINDOW_SECONDS) {
                fastAnswersCount++
            }
        }

        return AnswerValidationResult.Valid(scoreResult)
    }

    val averageResponseTimeSeconds: Double
        get() = if (responseTimes.isEmpty()) 0.0 else responseTimes.average()

    val accuracyPercentage: Double
        get() = if (totalAnswered == 0) 0.0 else (correctCount.toDouble() / totalAnswered * 100.0)
}

File: src/core/challenge/ServerScoringValidator.ts
import { Question } from '../../data/models/Question';
import { AnswerSubmissionEvent } from '../../data/models/ChallengeModels';
import { QuestionScoreResult } from '../types';
import { ScoringEngine } from '../scoring/ScoringEngine';

export type AntiCheatViolation = 
  | 'CHALLENGE_EXPIRED'
  | 'DUPLICATE_QUESTION_SUBMISSION'
  | 'SEQUENCE_MISMATCH'
  | 'INVALID_QUESTION_ID'
  | 'IMPOSSIBLE_RESPONSE_TIME'
  | 'REPLAY_DETECTED';

export type AnswerValidationResult = 
  | { status: 'VALID'; scoreResult: QuestionScoreResult }
  | { status: 'REJECTED'; reason: AntiCheatViolation };

export class ServerScoringValidator {
  public readonly challengeId: string;
  public readonly startTimeMillis: number;
  public readonly endTimeMillis: number;
  private readonly officialSequence: Question[];

  private answeredQuestionIds: Set<string> = new Set();
  private processedEventIds: Set<string> = new Set();
  private responseTimes: number[] = [];

  public verifiedScore: number = 0.0;
  public currentCombo: number = 0;
  public bestCombo: number = 0;
  public totalAnswered: number = 0;
  public correctCount: number = 0;
  public fastAnswersCount: number = 0;

  constructor(
    challengeId: string,
    startTimeMillis: number,
    endTimeMillis: number,
    officialSequence: Question[]
  ) {
    this.challengeId = challengeId;
    this.startTimeMillis = startTimeMillis;
    this.endTimeMillis = endTimeMillis;
    this.officialSequence = officialSequence;
  }

  public validateAndScore(
    event: AnswerSubmissionEvent,
    serverTimestamp: number = Date.now()
  ): AnswerValidationResult {
    // 1. Replay attack check
    if (this.processedEventIds.has(event.eventId)) {
      return { status: 'REJECTED', reason: 'REPLAY_DETECTED' };
    }
    this.processedEventIds.add(event.eventId);

    // 2. Challenge expiration check
    if (serverTimestamp > this.endTimeMillis + 2000) { // 2s grace for network transit
      return { status: 'REJECTED', reason: 'CHALLENGE_EXPIRED' };
    }

    // 3. Duplicate question check
    if (this.answeredQuestionIds.has(event.questionId)) {
      return { status: 'REJECTED', reason: 'DUPLICATE_QUESTION_SUBMISSION' };
    }

    // 4. Sequence mismatch check
    if (event.sequencePosition < 0 || event.sequencePosition >= this.officialSequence.length) {
      return { status: 'REJECTED', reason: 'SEQUENCE_MISMATCH' };
    }
    const expected = this.officialSequence[event.sequencePosition];
    if (expected.questionId !== event.questionId) {
      return { status: 'REJECTED', reason: 'SEQUENCE_MISMATCH' };
    }

    // 5. Impossible reflex time (< 50ms)
    if (event.responseTimeMs < 50) {
      return { status: 'REJECTED', reason: 'IMPOSSIBLE_RESPONSE_TIME' };
    }

    this.answeredQuestionIds.add(event.questionId);
    this.totalAnswered += 1;

    const responseSeconds = event.responseTimeMs / 1000.0;
    this.responseTimes.push(responseSeconds);

    const isCorrect = event.selectedAnswer === expected.correctAnswer;
    const scoreResult = ScoringEngine.scoreAnswer(
      isCorrect,
      responseSeconds,
      this.currentCombo,
      this.bestCombo
    );

    this.verifiedScore = ScoringEngine.roundToOneDecimal(this.verifiedScore + scoreResult.totalQuestionScore);
    this.currentCombo = scoreResult.currentCombo;
    this.bestCombo = scoreResult.bestCombo;

    if (isCorrect) {
      this.correctCount += 1;
      if (responseSeconds <= ScoringEngine.SPEED_WINDOW_SECONDS) {
        this.fastAnswersCount += 1;
      }
    }

    return { status: 'VALID', scoreResult };
  }

  public get averageResponseTimeSeconds(): number {
    if (this.responseTimes.length === 0) return 0.0;
    const sum = this.responseTimes.reduce((a, b) => a + b, 0);
    return ScoringEngine.roundToOneDecimal(sum / this.responseTimes.length);
  }

  public get accuracyPercentage(): number {
    if (this.totalAnswered === 0) return 0.0;
    return ScoringEngine.roundToOneDecimal((this.correctCount / this.totalAnswered) * 100.0);
  }
}
